File: Stack/05.EvaluateReversePolishNotation/evaluate_reverse_polish_notation.go
package main

import "fmt"
import "strconv"

type Stack struct {
	nums []int
}

func (self *Stack) push(num int) {
	self.nums = append(self.nums, num)
}

func (self *Stack) pop() int {
	top := self.nums[len(self.nums)-1]
	self.nums = self.nums[:len(self.nums)-1]
	return top
}

func evalRPN(tokens []string) int {

	if len(tokens) == 0 {
		panic("no num, no operations")
	}

	operations := map[string]bool{
		"+": true,
		"-": true,
		"*": true,
		"/": true,
	}

	//handle one num case
	if len(tokens) == 1 && operations[tokens[0]] == false {
		result, err := strconv.Atoi(tokens[0])
		if err == nil {
			return result
		}
	}

	stack := &Stack{}
	result := 0
	for _, token := range tokens {
		if operations[token] {
			result = calculate(token, stack.pop(), stack.pop())
			stack.push(result)
		} else {
			num, err := strconv.Atoi(token)
			if err == nil {
				stack.push(num)
			}
		}
	}

	return result
}

func calculate(operation string, a int, b int) int {
	switch operation {
	case "+":
		return b + a
	case "-":
		return b - a
	case "*":
		return b * a
	case "/":
		return b / a
	}

	panic("illegal oprator")
}

func main() {
	fmt.Println(evalRPN([]string{"18"}))
	fmt.Println(evalRPN([]string{"2", "1", "+", "3", "*"}))
	fmt.Println(evalRPN([]string{"4", "13", "5", "/", "+"}))
	fmt.Println(evalRPN([]string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}))
}

File: Stack/05.EvaluateReversePolishNotation/evaluateReversePolishNotation.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> myStack;
        for(auto str : tokens) {
            if(str != "+" && str != "-" && str != "*" && str != "/") {
                myStack.push(stoi(str));
            }
            else {
                int second = myStack.top(); myStack.pop();
                int first = myStack.top(); myStack.pop();
                int result;
                switch (str[0]) {
                    case '+':{
                        result = first + second;
                        break;
                    }
                    case '-':{
                        result = first - second;
                        break;
                    }
                    case '*':{
                        result = first * second;
                        break;
                    }
                    case '/':{
                        result = first / second;
                        break;
                    }
                    default:{
                        result = 0;
                        break;
                    }
                }
                myStack.push(result);
            }
        }
        return myStack.size() == 1 ? myStack.top() : 0;
    }
};